### ISSUE TITLE
XDS state machine issue with SDS request being ignored

### ISSUE THREAD

**Issue by shashankram** [openservicemesh/osm#3775]
**Bug description**:
There seems to be a bug in osm-controller's XDS state machine where requested resources are not correctly fulfilled, potentially due to how the requested resources are tracked.

In the demo, it was observed that 2 secrets on the bookbuyer proxy were not fulfilled:
```
  {
   "@type": "type.googleapis.com/envoy.admin.v3.SecretsConfigDump",
   "dynamic_warming_secrets": [
    {
     "name": "service-cert:bookbuyer/bookbuyer",
     "version_info": "uninitialized",
     "last_updated": "2021-07-16T21:11:56.339Z",
     "secret": {
      "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret",
      "name": "service-cert:bookbuyer/bookbuyer"
     }
    },
    {
     "name": "root-cert-for-mtls-outbound:bookstore/bookstore",
     "version_info": "uninitialized",
     "last_updated": "2021-07-16T21:11:56.339Z",
     "secret": {
      "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret",
      "name": "root-cert-for-mtls-outbound:bookstore/bookstore"
     }
    }
   ]
  }
```

OSM controller logs: 
[osm_controller.log](https://github.com/openservicemesh/osm/files/6833594/osm_controller.log)

OSM controller logs related to bookbuyer proxy:
[osm_controller_bookbuyer.log](https://github.com/openservicemesh/osm/files/6833595/osm_controller_bookbuyer.log)

Config dump on bookbuyer:
[bookbuyer_envoy_dump.txt](https://github.com/openservicemesh/osm/files/6833605/bookbuyer_envoy_dump.txt)

This seemed to happen when switching from permissive mode to SMI. From the logs, it seems as though the controller thinks the 2 requested SDS secrets are ACK'd, when in reality this is a request from Envoy post the LDS/CDS config update after switching from permissive mode to SMI:
```
 85 {"level":"debug","component":"envoy/ads","time":"2021-07-16T21:11:37Z","file":"stream.go:1    80","message":"Proxy Proxy: [Serial: 257302841095315423454589572387129624217], [Pod metada    ta: UID=7a455a42-ff65-41aa-8175-97faf6ad0707, Namespace=bookbuyer, Name=bookbuyer-5f5869b6    b8-t57sn, ServiceAccount=bookbuyer]: Request type.googleapis.com/envoy.extensions.transpor    t_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[root-cert-for-mt    ls-outbound:bookstore/bookstore service-cert:bookbuyer/bookbuyer]] last sent [nonce=162646    9872037692003; version=3]"}
 86 {"level":"debug","component":"envoy/ads","time":"2021-07-16T21:11:37Z","file":"stream.go:2    78","message":"Proxy Proxy: [Serial: 257302841095315423454589572387129624217], [Pod metada    ta: UID=7a455a42-ff65-41aa-8175-97faf6ad0707, Namespace=bookbuyer, Name=bookbuyer-5f5869b6    b8-t57sn, ServiceAccount=bookbuyer]: ACK received for SDS, version: 3 nonce: 1626469872037    692003 resources ACKd: Set{root-cert-for-mtls-outbound:bookstore/bookstore, service-cert:b    ookbuyer/bookbuyer}"}
 87 {"level":"debug","component":"envoy/ads","time":"2021-07-16T21:11:37Z","file":"stream.go:1    80","message":"Proxy Proxy: [Serial: 257302841095315423454589572387129624217], [Pod metada    ta: UID=7a455a42-ff65-41aa-8175-97faf6ad0707, Namespace=bookbuyer, Name=bookbuyer-5f5869b6    b8-t57sn, ServiceAccount=bookbuyer]: Request type.googleapis.com/envoy.extensions.transpor    t_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[service-cert:boo    kbuyer/bookbuyer]] last sent [nonce=1626469872037692003; version=3]"}
 88 {"level":"debug","component":"envoy/ads","time":"2021-07-16T21:11:37Z","file":"stream.go:2    78","message":"Proxy Proxy: [Serial: 257302841095315423454589572387129624217], [Pod metada    ta: UID=7a455a42-ff65-41aa-8175-97faf6ad0707, Namespace=bookbuyer, Name=bookbuyer-5f5869b6    b8-t57sn, ServiceAccount=bookbuyer]: ACK received for SDS, version: 3 nonce: 1626469872037    692003 resources ACKd: Set{service-cert:bookbuyer/bookbuyer}"}
--
The log in L88 indicates this request was not responded to, since it was treated as an ACK rather than a DiscoveryRequest
```

**Affected area (please mark with X where applicable)**:
- Envoy Control Plane        [X]

**Expected behavior**:
The secrets should be provisioned correctly, based on the logs it seems there is some sort of issue in the XDS state machine where this request doesn't come to SDS.

**Comment by eduser25** [comment]
taking a look

**Comment by eduser25** [comment]
@shashankram thanks for filing, it is indeed as you say. Leaving a bit of the problem decumented here.

The tl;dr I think we have not implemented unsubscription properly (if at all), though the exact scenario we are seeing is not entirely described in [XDS's protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#unsubscribing-from-resources):

> In the SotW protocol variants, each request must contain the full list of resource names being subscribed to in the resource_names field, so unsubscribing to a set of resources is done by sending a new request containing all resource names that are still being subscribed to but not containing the resource names being unsubscribed to. For example, if the client had previously been subscribed to resources A and B but wishes to unsubscribe from B, it must send a new request containing only resource A.

However it does not specifically say if the new request requires to be responded with a new response if the previous response already fulfills it fully and entirely (I suspect we have to, after debugging this issue). 
Neither [which resources to send](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return) section nor [when to send an update](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#when-to-send-an-update) touch directly this scenario. 

---

During boot phase, OSM eventually reply the the last boot request for envoy, and we properly send it the right certs:
```
Empty nonce for SDS, should be first message on stream (req resources: [root-cert-for-mtls-outbound:bookstore/bookstore root-cert-for-mtls-outbound:bookwarehouse/bookwarehouse service-cert:bookbuyer/bookbuyer])"}
...
requested cert root-cert-for-mtls-outbound:bookstore/bookstore"}
requested cert root-cert-for-mtls-outbound:bookwarehouse/bookwarehouse"}
requested cert service-cert:bookbuyer/bookbuyer"}
```
`1626469872037692003` nonce is created and sent with the above certs as a response. So nonce `003` (let's call it that) was sent to envoy with the tree requested certs.

After a few other ignored requests on older nonces, eventually envoy catches up with `003`:
```
ACK received for SDS, version: 3 nonce: 1626469872037692003 resources ACKd: Set{root-cert-for-mtls-outbound:bookstore/bookstore, root-cert-for-mtls-outbound:bookwarehouse/bookwarehouse, service-cert:bookbuyer/bookbuyer}"}
```
And it's a proper ACK as nonces, versions and resources match. So far so good.


Checking the times, the change from permissive to SMI happen around 15~20 secs after. The problem sort of starts to manifest just right after:
```
// Request from envoy with One lest cert, still acknowledging nonce 003
Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[root-cert-for-mtls-outbound:bookstore/bookstore service-cert:bookbuyer/bookbuyer]] last sent [nonce=1626469872037692003; version=3]"}
// Our logic thinks: you are on the right nonce, and the subset of certs requested are already present on that nonce, therefore ACK
ACK received for SDS, version: 3 nonce: 1626469872037692003 resources ACKd: Set{root-cert-for-mtls-outbound:bookstore/bookstore, service-cert:bookbuyer/bookbuyer}"}
// Request from envoy with One lest cert, still acknowledging nonce 003
Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[service-cert:bookbuyer/bookbuyer]] last sent [nonce=1626469872037692003; version=3]"}
// Our logic thinks: you are on the right nonce, and the subset of certs requested are already present on that nonce, therefore ACK
ACK received for SDS, version: 3 nonce: 1626469872037692003 resources ACKd: Set{service-cert:bookbuyer/bookbuyer}"}
// Request from envoy with NO certs, still acknowledging nonce 003
Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[]] last sent [nonce=1626469872037692003; version=3]"}
// What can we say.... "anything" that we might have sent you will match "nothing"
ACK received for SDS, version: 3 nonce: 1626469872037692003 resources ACKd: Set{}"}
```

The previous logs are basically unsubscription for all resources on SDS, one by one.
What happens later:

```
Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[service-cert:bookbuyer/bookbuyer]] last sent [nonce=1626469872037692003; version=3]"}
ACK received for SDS, version: 3 nonce: 1626469872037692003 resources ACKd: Set{service-cert:bookbuyer/bookbuyer}"}
Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1626469872037692003; version=3; resources=[root-cert-for-mtls-outbound:bookstore/bookstore service-cert:bookbuyer/bookbuyer]] last sent [nonce=1626469872037692003; version=3]"}
ACK received for SDS, version: 3 nonce: 1626469872037692003 resources ACKd: Set{root-cert-for-mtls-outbound:bookstore/bookstore, service-cert:bookbuyer/bookbuyer}"}
```

Since envoy is still referring to nonce `003`, and nonce `003` contained all the resources being asked, we treat everything as an ACK. Our logic interprets that "with the last nonce, you should be ok already", which I think is wrong in terms of unsubscription. 

I'd propose the following solution, I'll need some time to verify this:
1 - Do not compute `Difference` anymore between requested and last responded, and instead expect them to be `Equal` for it to be an ACK. This will ensure we respond to unsubscription requests with the exact resources requested.
2 - Record as `sentResources` only resources sent that are actually present on requested. We are currently not doing this and this could lead to the same set of issues if we ever respond something more than expected and we record it ahead of time.

**Comment by eduser25** [comment]
We are done here, both added e2e and validated with @trstringer the use case that showed the issue.
